zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -0,0 +1,181 @@
+//! Adds Sapling and Orchard note commitment subtrees to database
+
+use std::sync::{
+    atomic::{self, AtomicBool},
+    mpsc, Arc,
+};
+
+use zebra_chain::{block::Height, parallel::tree::TRACKED_SUBTREE_HEIGHT, parameters::Network};
+
+use crate::{service::finalized_state::ZebraDb, DiskWriteBatch};
+
+use super::DiskFormatUpgrade;
+
+/// Implements [`DiskFormatUpgrade`] for adding Sapling and Orchard note commitment subtrees to database
+pub struct AddSubtrees;
+
+impl DiskFormatUpgrade for AddSubtrees {
+    fn run(
+        &self,
+        initial_tip_height: Height,
+        _network: Network,
+        upgrade_db: &ZebraDb,
+        should_cancel_format_change: Arc<AtomicBool>,
+    ) {
+        let Some((sapling_subtree_start_height, _)) = upgrade_db
+            .sapling_tree_by_height_range(Height(1)..)
+            .next() else { return };
+
+        let orchard_subtree_start_height = upgrade_db
+            .orchard_tree_by_height_range(Height(1)..)
+            .next()
+            .map(|(h, _)| h);
+
+        // Create an unbounded channel for reading note commitment trees
+        let (sapling_tree_tx, sapling_tree_rx) = mpsc::channel();
+
+        // Read sapling note commitment trees
+        let db = upgrade_db.clone();
+        let should_cancel_flag = should_cancel_format_change.clone();
+        let sapling_read_task = std::thread::spawn(move || {
+            let read_from = sapling_subtree_start_height.next();
+            for (height, tree) in db.sapling_tree_by_height_range(read_from..initial_tip_height) {
+                // Breaking from this loop and dropping the sapling_tree channel
+                // will cause the sapling compare and delete tasks to finish.
+                if should_cancel_flag.load(atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Err(error) = sapling_tree_tx.send((height, tree)) {
+                    panic!("unexpected send error: {error:?}");
+                }
+            }
+        });
+
+        // Create an unbounded channel for sending sapling note commitment subtrees to be written
+        let (sapling_subtree_tx, sapling_subtree_rx) = mpsc::channel();
```

Comment:
Overall, I feel like this design is more complicated than it needs to be. Is it necessary to launch separate threads to do this upgrade?

It could be easier to implement it without threads, and then see if the performance is acceptable. (Remember, the upgrade task itself is already running in a separate thread.)

#### Security / Reliability

Using an unbounded channel can potentially read all the trees into RAM, if the write thread happens to block. How much memory is that? If it's a large amount, this is a potential hang or crash issue.

(I think it should be ok after duplicates are removed, and the duplicate removal task itself is ok because its compare task drops most of those trees immediately.)